routemaster/state_machine/api.py

```diff
@@ -113,186 +89,147 @@ def create_label(app: App, label: Label, metadata: Metadata) -> Metadata:
+            metadata_triggers_processed=not needs_gate_evaluation,
+        ))
 
-def delete_label(app: App, label: Label) -> None:
+    # Outside transaction
+    if needs_gate_evaluation:
+        try:
+            with app.db.begin() as conn:
+                lock_label(label, conn)
+                current_state = get_current_state(label, state_machine, conn)
+                if not isinstance(current_state, Gate):  # pragma: no branch
+                    # Cannot be hit because of the semantics of
+                    # `needs_gate_evaluation_for_metadata_change`. Here to
+                    # appease mypy.
+                    raise TypeError("Label not in a gate")  # pragma: no cover
+                could_progress = process_gate(app, current_state, label, conn)
+
+            if could_progress:
+                _process_transitions(app, label)
+        except Exception:
+            # This is allowed to fail here. We have successfully saved the new
+            # metadata, and it has a metadata_triggers_processed=False flag so
+            # will be picked up again for processing later.
+            pass
+
+    return new_metadata
+
+
+def _process_transitions(app: App, label: LabelRef):
+    state_machine = get_state_machine(app, label)
+    could_progress = True
+
+    while could_progress:
+        with app.db.begin() as conn:
+            lock_label(label, conn)
+            current_state = get_current_state(label, state_machine, conn)
+
+            if isinstance(current_state, Action):
+                could_progress = process_action(
+                    app,
+                    current_state,
+                    label,
+                    conn,
+                )
+
+            elif isinstance(current_state, Gate):  # pragma: no branch
+                if not current_state.trigger_on_entry:
+                    return
+
+                could_progress = process_gate(
+                    app,
+                    current_state,
+                    label,
+                    conn,
+                )
+
+            else:
+                raise RuntimeError(  # pragma: no cover
+                    "Unsupported state type {0}".format(current_state),
+                )
+
+
+def delete_label(app: App, label: LabelRef) -> None:
     """
     Deletes the metadata for a label and marks the label as deleted.
 
     The history for the label is not changed (in order to allow post-hoc
     analysis of the path the label took through the state machine).
     """
-    state_machine = _get_state_machine(app, label)
+    state_machine = get_state_machine(app, label)
 
-    metadata_field = labels.c.metadata
     label_filter = and_(
         labels.c.name == label.name,
         labels.c.state_machine == state_machine.name,
     )
 
     with app.db.begin() as conn:
         existing_metadata = conn.scalar(
-            select([metadata_field]).where(label_filter),
+            select([labels.c.metadata]).where(label_filter),
```

Comment:
I think we should be locking the label here too. Not sure whether that's this PR or not, but wanted to record that we should be doing it.

Reply:
This is now done and pushed.